docs/howtos/model_surgery.rst

```diff
@@ -128,105 +126,67 @@ After doing whatever you want, unflatten back:
 
 .. testcode::
 
-    # Unflatten
-    unflat_opt_state = traverse_util.unflatten_dict({tuple(k.split('/')): v for k, v in flat_opt_state.items()})
-    print(jax.tree_map(jnp.shape, unflat_opt_state))
+  flat_mu = traverse_util.flatten_dict(opt_state[0].mu, sep='/')
+  flat_nu = traverse_util.flatten_dict(opt_state[0].nu, sep='/')
+
+  print(jax.tree_map(jnp.shape, flat_mu))
 
 .. testoutput::
   :options: +NORMALIZE_WHITESPACE
   
-  {'state': {'param_states': {'Conv_0': {'bias': {'grad_ema': (32,),
-      'grad_sq_ema': (32,)},
-      'kernel': {'grad_ema': (3, 3, 1, 32), 'grad_sq_ema': (3, 3, 1, 32)}},
-    'Conv_1': {'bias': {'grad_ema': (64,), 'grad_sq_ema': (64,)},
-      'kernel': {'grad_ema': (3, 3, 32, 64), 'grad_sq_ema': (3, 3, 32, 64)}},
-    'Dense_0': {'bias': {'grad_ema': (256,), 'grad_sq_ema': (256,)},
-      'kernel': {'grad_ema': (3136, 256), 'grad_sq_ema': (3136, 256)}},
-    'Dense_1': {'bias': {'grad_ema': (10,), 'grad_sq_ema': (10,)},
-      'kernel': {'grad_ema': (256, 10), 'grad_sq_ema': (256, 10)}}},
-    'step': ()},
-  'target': {'Conv_0': {'bias': (32,), 'kernel': (3, 3, 1, 32)},
-    'Conv_1': {'bias': (64,), 'kernel': (3, 3, 32, 64)},
-    'Dense_0': {'bias': (256,), 'kernel': (3136, 256)},
-    'Dense_1': {'bias': (10,), 'kernel': (256, 10)}}}
-
-We can restore the optimizer object from the nested-dict state. The restored 
-state must agree with the shape of the existing object as a sort of "structural
-unit test".
-
-.. testcode::
+  {'Conv_0/bias': (32,),
+   'Conv_0/kernel': (3, 3, 1, 32),
+   'Conv_1/bias': (64,),
+   'Conv_1/kernel': (3, 3, 32, 64),
+   'Dense_0/bias': (256,),
+   'Dense_0/kernel': (3136, 256),
+   'Dense_1/bias': (10,),
+   'Dense_1/kernel': (256, 10)}
 
-  restored_opt = opt.restore_state(unflat_opt_state)
-  print(jax.tree_map(jnp.shape, restored_opt))
+After modification, re-create optimizer state:
 
-.. testoutput::
-  :options: +NORMALIZE_WHITESPACE, +ELLIPSIS
+.. testcode::
 
-  Optimizer(optimizer_def=<flax.optim.adam.Adam object at ...>, state=OptimizerState(step=(), param_states={'Conv_0': {'bias': _AdamParamState(grad_ema=(32,), grad_sq_ema=(32,)), 'kernel': _AdamParamState(grad_ema=(3, 3, 1, 32), grad_sq_ema=(3, 3, 1, 32))}, 'Conv_1': {'bias': _AdamParamState(grad_ema=(64,), grad_sq_ema=(64,)), 'kernel': _AdamParamState(grad_ema=(3, 3, 32, 64), grad_sq_ema=(3, 3, 32, 64))}, 'Dense_0': {'bias': _AdamParamState(grad_ema=(256,), grad_sq_ema=(256,)), 'kernel': _AdamParamState(grad_ema=(3136, 256), grad_sq_ema=(3136, 256))}, 'Dense_1': {'bias': _AdamParamState(grad_ema=(10,), grad_sq_ema=(10,)), 'kernel': _AdamParamState(grad_ema=(256, 10), grad_sq_ema=(256, 10))}}), target={'Conv_0': {'bias': (32,), 'kernel': (3, 3, 1, 32)}, 'Conv_1': {'bias': (64,), 'kernel': (3, 3, 32, 64)}, 'Dense_0': {'bias': (256,), 'kernel': (3136, 256)}, 'Dense_1': {'bias': (10,), 'kernel': (256, 10)}})
+  opt_state = (
+      opt_state[0]._replace(
```

Comment:
I don't think we should be calling methods that begin with underscores from our formal documentation. What's the recommended Optax way to do this?

Reply:
Oh this is just a `collections.namedtuple` convention to avoid conflicts with field names:
https://docs.python.org/3/library/collections.html#collections.namedtuple

Reply:
Hmm though still it seems sketchy to explicitly ask people to use private methods. I don't know what the counterproposal would be but this doesn't seem right to me...